### PR TITLE
Issue 2066: clip trim snapping

### PR DESCRIPTION
**Pull request by vsverchinsky**
Resolves: #2066 

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h:45]: One of the C++ Core Guidelines I like is that each virtual function declaration should be marked with exactly one of `virtual` `override` or `final` as appropriate, and for these two it should be `override`.

Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h:45]: Oh, I see I misunderstood the context.  This is not overriding UIHandle functions.  `virtual` is right.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:101]: Clang gives me a warning about '&&' within '||'.  Add parentheses to make the intent unmistakable.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:103]: Same track, or aligned partner track

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:104]: I would prefer to iterate the intervals, as in the other branch, not the clips, for uniformity, and in case WaveTrack might add some other kind of interval in future (who knows?)

You can static_cast the Extra() pointer of each interval to WaveTrack::IntervalData* and then test for the clip you want to exclude.

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:130]: The two shared_ptr arguments could be const references, and so they should, to aid the human reader.  Instead I read this then stopped to see whether `track` and `clip` were being reassigned for some reason.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:157]: A const auto & here for consistency

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:389]: Another argument might be const reference

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h:87]: See how SelectHandle overrides Enter, HasEscape, and Escape member functions.  These allow you to hit the Esc key and dismiss the yellow snap lines, enabling a fine, un-snapped adjustment, if that is what the user really wants.  The same might be done here.

Maybe some common base class, derived from UIHandle, should be defined?  Hm.

Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h:87]: You may note that I could have done the extra work I suggested in the UIHandle subclasses recently added to AdornedRulerPanel, but I didn't.

This is not mandatory extra work.  But I should reconsider how the pattern might be made easier and reusable across UIHandle classes that snap.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:384]: See TrackPanelDrawable::DrawingArea:  default behavior should not be to return an empty wxRect, but to return the second argument unchanged.

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (vsverchinsky) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp]: Funny, the tool I use to analyze a code suggests quite opposite: it says that parenthesizes are redundant :) But I usually tend to put them too. Thank you!

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (vsverchinsky) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp]: Didn't think of that. Thank you!

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h
Comment (vsverchinsky) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h:87]: > See how SelectHandle overrides Enter, HasEscape, and Escape member functions. These allow you to hit the Esc key and dismiss the yellow snap lines, enabling a fine, un-snapped adjustment, if that is what the user really wants. The same might be done here.
> 
> Maybe some common base class, derived from UIHandle, should be defined? Hm.

If I try to hit escape when using a select handle it just resets the current selection, with a debugger attached I see that both `::Escape` and `::Cancel` member functions are triggered, did I misunderstood something?

Comment (Paul-Licameli) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h:87]: You can hit escape to dismiss the yellow snap line before the click.  It seems you can't dismiss the yellow snap line while dragging.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:106]: Just a discussion...

I would prefer `static_cast` myself here...  But actually, I'd rather see `std::any_cast`.  The Extra information in IntervalData looks like a place I could rewrite things with `std::any` now that we use C++17.  I think `any_cast` makes a simpler check than `dynamic_cast`.  Just equality of a statically determined `typeid` of the template parameter with the dynamically determined `typeid` in the `any`.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:380]: Still not resolved

**Review by Paul-Licameli — CHANGES_REQUESTED** [review]

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h
Comment (vsverchinsky) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.h:87]: Does not seem to work as well, I still get snap lines during the selection

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (crsib) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:106]: I've never seen much performance impact from using `dynamic_cast`. It can be slow in tight loops with deep hierarchies, but that's not the case here.

Comment (vsverchinsky) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:106]: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Pro-type-downcast

I agree that sometimes we can justify using `static_cast` instead of `dynamic_cast` if we want to get as much performance as possible, and we do guarantee that the specified type is the only possible option here. But one of the reasons why this part of the code was rewritten is that we may want to add more interval types to a wave track in the future, wasn't it?

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (Paul-Licameli) [inline src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:179]: A general comment: I'm reviewing changes in the class AdjustBorders.  The other subclass of ClipTrimPolicy is AdjustBetweenBorders.  That class is not constructed anywhere.  Is there a mistake?

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (vsverchinsky) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:380]: Sorry, will fix that, thanks!

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (vsverchinsky) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:179]: It's not a mistake, AdjustBetweenBorders moves borders of both clips if they are close enough, and it was my understanding of how things should happen, and turned out that I was wrong. But I was told that we may need this in the future.

**Review by Paul-Licameli — APPROVED** [review]
Youl will fix the small matter about the rectangle, and then there is nothing else to block this.

File: src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp
Comment (crsib) [reply src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp:106]: `std::any` has costs attached. It's not a `zero-cost` abstraction in general, although compilers may implement a fast-pass for small objects.

**Comment by Paul-Licameli** [comment]
Please rebase it once more, then create a merge commit.

**Comment by vsverchinsky** [comment]
> Please rebase it once more, then create a merge commit.

Rebased, but I need an approve for this PR to merge

**Comment by Paul-Licameli** [comment]
> > Please rebase it once more, then create a merge commit.
> 
> Rebased, but I need an approve for this PR to merge

I'm sorry, I thought I had done that already.

**Review by Paul-Licameli — APPROVED** [review]
Approved!

**Comment by crsib** [comment]
> > > Please rebase it once more, then create a merge commit.
> > 
> > 
> > Rebased, but I need an approve for this PR to merge
> 
> I'm sorry, I thought I had done that already.

That's my fault! I will probably remove the "push invalidates review" rule from branch protection rules

**Comment by Paul-Licameli** [comment]
> > > > Please rebase it once more, then create a merge commit.
> > > 
> > > 
> > > Rebased, but I need an approve for this PR to merge
> > 
> > 
> > I'm sorry, I thought I had done that already.
> 
> That's my fault! I will probably remove the "push invalidates review" rule from branch protection rules

Was it that, or was it a dismissal of my stale review by Vitaly, above?

**Comment by crsib** [comment]
Vitaly pushed and that invalidated the review

**Pull request merged**
